Add rust search to the repository

llogiq opened the pull request:
This is a drop-in replacement for the python-based search service.

llogiq committed:
Add rust search to the repository

generall commented on rust_search/README.md:
````diff
@@ -0,0 +1,35 @@
+This version of the search uses a local model. So you need the model in the working path of the service. To get the ONNX file, run
+
+```bash
+curl -o all-MiniLM-L6-v2.onnx https://huggingface.co/optimum/all-MiniLM-L6-v2/blob/main/model.onnx
````
This command doesn't work

llogiq replied:
Oops, missed a /resolve in the path.

generall commented on rust_search/src/main.rs:
```diff
@@ -0,0 +1,293 @@
+    if let Some(Kind::StringValue(text)) = &payload.get("text").and_then(|v| v.kind.as_ref()) {
+        highlight(result, &take_n_chars(text, TEXT_LIMIT), q);
+    }
+    result.push_str("\"}");
```
Is there a reason to not use serde here? What if payload have un-escaped `"`?

generall commented on rust_search/src/main.rs:
```diff
@@ -0,0 +1,293 @@
+            .await
+        {
+            Ok(RecommendResponse { result, .. }) => result,
+            Err(e) => return Err(InternalError::new(e, StatusCode::BAD_GATEWAY)),
```
if `recommend` fails, it should re-try with regular search.

llogiq replied:
Shouldn't it re-try if it returns no result? If it *fails*, it's likely that the regular search will fail, too.

generall replied:
`recommend` will fail if there is no prefix id in the lookup collection

generall commented on rust_search/src/main.rs:
```diff
@@ -0,0 +1,293 @@
+            .await
+        {
+            Ok(SearchResponse { result, .. }) => result,
+            Err(e) => return Err(InternalError::new(e, StatusCode::BAD_GATEWAY)),
```
I don't think `BAD_GATEWAY` is correct error code in this case

llogiq replied:
Changed to INTERNAL_SERVER_ERROR.

generall commented on rust_search/src/main.rs:
```diff
@@ -0,0 +1,293 @@
+    result.push_str("\"}");
+}
+
+fn prefix_to_id(prefix: &str) -> PointId {
```
```
error[E0412]: cannot find type `PointId` in this scope
   --> src/main.rs:113:34
    |
113 | fn prefix_to_id(prefix: &str) -> PointId {
    |                                  ^^^^^^^ not found in this scope
    |
help: consider importing this struct
    |
1   + use qdrant_client::qdrant::PointId;
    |

```

llogiq replied:
Oops, missed the import.

generall commented on rust_search/src/main.rs:
```diff
@@ -0,0 +1,293 @@
+                }
+                _ => {}
+            }
+        }
```
we usually deploy behind the nginx/caddy service, it does proper certificate rotation e.t.c.. I would remove it here just to have less code to maintain.

generall commented on rust_search/src/main.rs:
```diff
@@ -0,0 +1,293 @@
+        }
+    }
+    result.push(']');
+    Ok(result)
```
Json response should have `Content-Type: application/json` header

generall commented on rust_search/src/main.rs:
```diff
@@ -0,0 +1,293 @@
+    let (tokenizer, session, qdrant) = context.get_ref();
+    let mut result = String::from("[");
+    let points = if q.len() < 5 {
+        let mut must = vec![Condition::matches("text", q.clone())];
```
this condition will exclude everything except for exact text match.

it should be

```suggestion
        let mut must = vec![Condition::matches("text", MatchValue::Text(q.clone()))];
```

generall replied:
Actually, the original code is a bit complicated. It falls-back to search without the filter. It can be implemented as batch request for speed

llogiq replied:
This is no longer needed because the text is now stored in the ID to reduce search overhead.

generall commented on rust_search/src/main.rs:
```diff
@@ -0,0 +1,293 @@
+
+// Ignored as writing to a `String` is always Ok
+#[allow(unused_must_use)]
+fn write_value(result: &mut String, value: &Value) {
+    match &value.kind {
+        Some(Kind::DoubleValue(v)) => {
+            write!(result, "{}", v);
+        }
+        Some(Kind::IntegerValue(v)) => {
+            write!(result, "{}", v);
+        }
+        Some(Kind::StringValue(v)) => {
+            write!(result, "\"{}\"", v);
+        }
+        Some(Kind::BoolValue(v)) => {
+            write!(result, "{}", v);
+        }
+        Some(Kind::StructValue(Struct { fields })) => write_map(result, fields),
+        Some(Kind::ListValue(ListValue { values })) => {
+            result.push('[');
+            let mut values_iter = values.into_iter();
+            if let Some(value) = values_iter.next() {
+                write_value(result, value);
+                for value in values_iter {
+                    result.push(',');
+                    write_value(result, value);
+                }
+            }
+            result.push(']');
+        }
+        _ => result.push_str("null"),
+    }
+}
```
__serde__

llogiq committed:
address @generall's remarks

llogiq commented:
@generall Thanks for the review. I addressed your suggestions. Only one thing: If recommend *fails*, I still return internal server error. The only case I can see where the search would succeed if the recommend fails is if the prefix cache is not set up, and I don't want this to go unnoticed.

llogiq committed:
fix recommend filter

Commit:
make recommend errors roll over to search

llogiq commented:
I must correct myself. What I had overlooked was that there may be queries that are not in the prefix cache which will return a Not Found error, so I changed the code as per your suggestion, @generall. With that said, it might be a good idea to expand the prefix cache to contain common search terms which are not in the word list (e.g. people might search for `io_uring`).

llogiq committed:
remove rustls, make clippy happy

generall committed:
review fixes

generall commented:
@llogiq, I took a liberty to make a few changes
